File: javascript/67. add-binary.js
/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */

const getReverseBinaryArray = str => str.split('').reverse().map(digit => parseInt(digit));

const addBinary = function(bin1, bin2) {
    const bstr1 = getReverseBinaryArray(bin1);
    const bstr2 = getReverseBinaryArray(bin2);

    const result = [];
    const endSize = Math.max(bstr1.length, bstr2.length);
    let i = 0;
    let carry = 0;

    for (i = 0; i < endSize; i++) {
        const digit1 = bstr1.length >= i + 1 ? bstr1[i] : 0;
        const digit2 = bstr2.length >= i + 1 ? bstr2[i] : 0;
        const sum = digit1 + digit2 + carry;

        if (sum === 3) {
            result.push(1);
            carry = 1;
        }
        else if (sum === 2) {
            result.push(0);
            carry = 1;
        }
        else {
            result.push(sum);
            carry = 0;
        }
    }
    if (carry)
        result.push(1);

    return result.map(digit => digit.toString()).reverse().join('');
};

const testcases = {
    'input': [["11", "1"], ["1010", "1011"]],
    'output': ["100", "10101"],
};

testcases['input'].forEach((args, idx) => {
    const sol = getReverseBinaryArray(...args);
    console.assert(
        sol === testcases['output'][idx],
        `Expected output ${testcases['output'][idx]} for input <${args}>, got ${sol}`
    );
    console.log();
});

File: cpp/67. add-binary.cpp
//
// Created by Rutvik Patel on 1/31/24.
//

#include "include.h"

// solution
class Solution {
public:
    std::string addBinary(std::string s1, std::string s2) {
        std::reverse(s1.begin(), s1.end());
        std::reverse(s2.begin(), s2.end());

        int maxSize = static_cast<int>(std::max(s1.size(), s2.size()));
        int carry = 0;
        std::string result;

        for (int idx = 0; idx < maxSize; ++idx) {
            int digit1 = idx < s1.size() ? s1[idx] - '0' : 0;
            int digit2 = idx < s2.size() ? s2[idx] - '0' : 0;
            int sum = digit1 + digit2 + carry;

            if (sum == 3) {
                result.push_back('1');
                carry = 1;
            } else if (sum == 2) {
                result.push_back('0');
                carry = 1;
            } else {
                result.push_back(static_cast<char>(sum + '0'));
                carry = 0;
            }
        }

        if (carry) result.push_back('1');

        std::reverse(result.begin(), result.end());

        return result;
    }
};

// main
int main(__attribute__((unused)) int argc, __attribute__((unused)) char *argv[]) {
    std::vector<std::pair<std::pair<std::string, std::string>, std::string>> test_cases = {
        { { "11", "1" }, "100"},
        { { "1010", "1011" }, "10101"},
    };

    Solution solution;

    for (auto& [args, ans] : test_cases)
        assert_m(solution.addBinary(args.first, args.second) == ans, "Failed!");

    log("All testcases passed!");
}